KAFKA-5484: Refactor kafkatest docker support

cmccabe opened the pull request.

cmccabe committed:
KAFKA-5484: Refactor kafkatest docker support

cmccabe commented:
This PR adds a lot of improvements for running kafka tests on Docker.

* Allow limiting the amount of memory which docker containers take
* Allow using iptables and sudo from inside docker
* Do not run tests as root -- run them as the ducker user instead
* set DEBIAN_FRONTEND in the Dockerfile
* Add the ability to log into docker containers from the shell (ducker-ak ssh)
* Add the ability to purge old docker containers that are no longer in use (ducker-ak purge)
* Fix an issue where docker containers were sometimes not rebuilt for users after the Dockerfile was updated
* Generate a new private key for each new Docker container, rather than using a static private key checked into Kafka

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5529/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5515/
Test PASSed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5560/
Test PASSed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5573/
Test PASSed (JDK 7 and Scala 2.11).

ijuma commented:
cc @ewencp

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5578/
Test FAILed (JDK 8 and Scala 2.12).

cmccabe commented:
retest this please

cmccabe commented:
This PR is needed to implement some of the fault injection stuff.  For now, I'll roll this PR into the follow-on PRs

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5584/
Test PASSed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5598/
Test PASSed (JDK 7 and Scala 2.11).

ewencp commented on tests/docker/Dockerfile:
```diff
@@ -19,15 +19,35 @@ MAINTAINER Apache Kafka dev@kafka.apache.org
+
+# Update Linux and install necessary utilities.
+RUN apt update && apt install -y sudo netcat iptables rsync unzip wget curl jq coreutils openssh-server net-tools vim python-pip python-dev libffi-dev libssl-dev && apt-get -y clean
+RUN pip install -U pip && pip install --upgrade cffi virtualenv pyasn1 && pip install --upgrade ducktape==0.6.0
```
Should we maybe add a note to https://github.com/apache/kafka/blob/trunk/tests/setup.py#L53 to note that we should also update the version here? Version updates (of Kafka, of ducktape) are starting to get scattered and hard to keep track of... unfortunately I don't know of a way to make this refer to some external file to load the version info from.

cmccabe replied:
Hmm, good point.  I added a comment on setup.py explaining this.

ewencp reviewed (commented):
One nit, but otherwise LGTM

cmccabe committed:
Add a comment to setup.py

Add a comment to setup.py explaining that when the ducktape version
changes the Dockerfile needs to change too.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5805/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5791/
Test FAILed (JDK 8 and Scala 2.12).

ewencp commented:
LGTM
